services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/internal/immutable/ImmutableIntrospector.java

```diff
@@ -0,0 +1,226 @@
+                            .propertyDescriptors(propertyDescriptors)
+                            .build();
+    }
+
+    private Optional<Method> findStaticBuilderMethod(Class<?> immutableClass, Class<?> builderClass) {
+        try {
+            Method method = immutableClass.getMethod(BUILDER_METHOD);
+
+            if (Modifier.isStatic(method.getModifiers()) && method.getReturnType().isAssignableFrom(builderClass)) {
+                return Optional.of(method);
+            }
+        } catch (NoSuchMethodException ignored) {
+            // no-op
+        }
+
+        return Optional.empty();
+    }
+
+    private IllegalArgumentException generateExceptionForMethod(Method getter, String message) {
+        return new IllegalArgumentException(
+            message + " Use the @DynamoDbIgnore annotation on the method if you do not want it to be included in the " +
+                "TableSchema introspection. [Method = \"" + getter + "\"]");
+    }
+
+    private Class<?> validateAndGetBuilderClass(Class<?> immutableClass) {
+        DynamoDbImmutable dynamoDbImmutable = immutableClass.getAnnotation(DynamoDbImmutable.class);
+
+        if (dynamoDbImmutable == null) {
+            throw new IllegalArgumentException("A DynamoDb immutable class must be annotated with @DynamoDbImmutable");
+        }
+
+        return dynamoDbImmutable.builder();
+    }
+
+    private void validateGetter(Method getter) {
+        if (getter.getReturnType() == void.class || getter.getReturnType() == Void.class) {
+            throw generateExceptionForMethod(getter, "A method was found on the immutable class that does not appear " +
+                "to be a valid getter due to the return type being void.");
+        }
+
+        if (getter.getParameterCount() != 0) {
+            throw generateExceptionForMethod(getter, "A method was found on the immutable class that does not appear " +
+                "to be a valid getter due to it having one or more parameters.");
+        }
+    }
+
+    private List<Method> filterAndCollectGetterMethods(Method[] rawMethods) {
+        return Arrays.stream(rawMethods)
+                     .filter(ImmutableIntrospector::isMappableMethod)
+                     .collect(Collectors.toList());
+    }
+
+    private Map<String, Method> filterAndIndexBuilderMethods(Method[] rawMethods) {
+        return Arrays.stream(rawMethods)
+                     .filter(ImmutableIntrospector::isMappableMethod)
+                     .collect(Collectors.toMap(this::normalizeSetterName, m -> m));
+    }
+
+    private String normalizeSetterName(Method setter) {
+        String setterName = setter.getName();
+
+        if (setterName.length() > 3
+            && Character.isUpperCase(setterName.charAt(3))
+            && setterName.startsWith(SET_PREFIX)) {
+
+            return Character.toLowerCase(setterName.charAt(3)) + setterName.substring(4);
+        }
+
+        return setterName;
+    }
```

Comment:
Should we make "is bean style" an annotation setting, instead of trying to guess by looking at the method names? I'm worried about an attribute starting with "Set" being magically messed up because we're trying to be too clever. 

E.g. A customer probably wants a "setOfStuff" attribute to be stored as "setOfStuff", not "ofStuff".

Reply:
Some code generators will produce immutable implementations that have 'set'. I think it would be inconvenient to require an additional annotation for all of those. I feel it would be preferable to annotate the exceptions with @DynamoDbAttribute("setOfStuff") (which is already supported)

Reply:
I was thinking a boolean on the class-level annotation, instead of one on every attribute. That seems less troublesome and prevents magic confusion (I feel like this is the opposite argument we usually have - you're usually the anti-magic one!)

Reply:
As we discussed offline, we'll bias towards making it easy for customers using code generators such as Lombok and not require an additional opt-in flag for getter/setter prefixes.